src/lib/logical_query_plan/abstract_lqp_node.hpp

```diff
@@ -73,7 +73,9 @@ class AbstractLQPNode : public std::enable_shared_from_this<AbstractLQPNode>, pr
   explicit AbstractLQPNode(LQPNodeType node_type);
 
   // Creates a deep copy
-  virtual std::shared_ptr<AbstractLQPNode> deep_copy() const;
+  using PreviousCopiesMap = std::map<std::shared_ptr<const AbstractLQPNode>, std::shared_ptr<AbstractLQPNode>>;
+  std::shared_ptr<AbstractLQPNode> deep_copy(
+      std::shared_ptr<PreviousCopiesMap> previous_copies = std::make_shared<PreviousCopiesMap>()) const;
```

Comment:
Ugh, having this in the public interface is ugly, even though it removes the to move all the deep_copy code into a `_deep_copy_impl(PreviousCopiesMap& map)` - with deep copy just creating the map and calling the impl method. Still, I'd prefer this more cumbersome approach that keeps the public interface clean.

Reply:
also, you could probably use unordered_map here? Your choice, I use it by default because probably speed(?)

Reply:
yepp